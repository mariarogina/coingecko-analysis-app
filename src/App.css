.App {
  text-align: center;
  font-size: x-large;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

.infoButton {
  margin: 10px;
  font-size: large;
  padding: 10px;
}
.appTitle {
  font-size: 2.4rem;
}
.inputElement {
  font-size: 1.2rem;
}

.submitButton {
  font-size: 1.2rem;
  padding: 8px;
}
.formItem {
  margin: 10px;
}

.appParagraph {
  margin: 50px;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
