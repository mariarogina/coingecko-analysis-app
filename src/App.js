import "./App.css";
import { useState } from "react";
import axios from "axios";
import {
  findLongestDownward,
  findMaxVolume,
  findMaxPrice,
  findMinPrice,
} from "./utils";

function App() {
  const [financialData, setFinancialData] = useState([]);
  const [startDate, setStartDate] = useState("");
  const [endDate, setEndDate] = useState("");
  const [totalVolumes, setTotalVolumes] = useState([]);
  const [maxVolume, setMaxVolume] = useState(0);
  const [downwardSpan, setDownwardSpan] = useState("");
  const [maxPriceDate, setMaxPriceDate] = useState("");
  const [minPriceDate, setMinPriceDate] = useState("");
  const [maxVolumeDate, setMaxVolumeDate] = useState("");
  const [canTrade, setCanTrade] = useState(true);
  const [showTrade, setShowTrade] = useState(false);
  const [showVolume, setShowVolume] = useState(false);
  const [showDown, setShowDown] = useState(false);
  const [dataLoaded, setDataLoaded] = useState(false);
  const [message, setMessage] = useState("");
  const [loadMessage, setLoadMessage] = useState("");

  const handleSubmit = async (e) => {
    e.preventDefault();
    async function fetchData() {
      try {
        const resp = await axios.get(
          `https://api.coingecko.com/api/v3/coins/bitcoin/market_chart/range?vs_currency=eur&from=${
            startDate + 3600
          }&to=${endDate}`
        );
        console.log(resp.data);
        setMessage("");
        setLoadMessage("");
        setFinancialData(filterEntries(resp.data.prices));
        setTotalVolumes(filterEntries(resp.data.total_volumes));
        console.log("DATA", financialData);
        console.log("VOLUMES", totalVolumes);
        setDataLoaded(true);
        setLoadMessage("The data has NOT loaded, try to enter some dates");
      } catch (err) {
        // Handle Error Here
        console.error(err);
        setMessage("Data not loaded, try entering some dates");
      }
    }
    fetchData();
  };

  const filterEntries = (list) => {
    const uniqueDays = [];
    const uniqueEntries = [];
    console.log("LIST TO FILTER:", list);
    list.forEach(function (item, index) {
      const dateDate = new Date(item[0]).toDateString();

      if (uniqueDays.indexOf(dateDate) < 0) {
        uniqueDays.push(dateDate);
        uniqueEntries.push(item);
      }
    });

    console.log("UNIQUE ENTRIES", uniqueEntries);
    return uniqueEntries;
  };

  const handleSubmitAndFilter = (e) => {
    setMessage("");
    setMaxVolume(0);
    setShowTrade(false);
    setDownwardSpan("");
    handleSubmit(e);
  };
  const convertDate = (date) => {
    const stamp = parseInt((new Date(date).getTime() / 1000).toFixed(0));
    return stamp;
  };

  const convertToISO = (stampDate) => {
    const jsDate = new Date(stampDate).toLocaleDateString("ru-RU");
    return jsDate;
  };

  const handleSetStart = (e) => {
    setStartDate("");
    setStartDate(convertDate(e.target.value));
  };
  const handleSetEnd = (e) => {
    setEndDate("");
    const newEndDate = convertDate(e.target.value);
    setEndDate(newEndDate);
  };

  //The maximum amount of days bitcoinâ€™s price was decreasing (dayspan)
  const handleFindDownward = () => {
    setShowDown(true);
    setShowVolume(false);
    setShowTrade(false);
    setMessage("");
    setMaxVolume(0);
    if (dataLoaded) {
      const down = findLongestDownward(financialData);
      setDownwardSpan(down);
    } else {
      setMessage("Please enter some dates first for showing the trend");
    }
  };

  //The date with the highest trading volume and the volume (max total volume and corresponding date)
  const handleFindVolume = () => {
    setMessage("");
    setShowDown(false);
    setShowTrade(false);
    setShowVolume(true);
    if (dataLoaded) {
      setDownwardSpan("");
      const maxVolInfo = findMaxVolume(totalVolumes);
      const maximumVolume = maxVolInfo[0];
      const maximumVolumeDate = maxVolInfo[1];
      setMaxVolume(maximumVolume);
      setMaxVolumeDate(maximumVolumeDate);
      console.log(maxVolumeDate);
      console.log(convertToISO(maxVolumeDate));
    } else {
      setMessage("Please enter some dates first for showing volume");
    }
  };

  const handleShouldBuyOrSell = () => {
    setMessage("");
    setShowTrade(true);
    setShowDown(false);
    setShowVolume(false);
    const impossible = ifOnlyDownward();
    if (dataLoaded) {
      if (impossible) {
        setCanTrade(false);
      } else {
        setCanTrade(true);
        handleFindSell();
      }
    } else {
      setMessage(
        "Please enter some dates first for showing buying and selling dates"
      );
    }
  };

  const ifOnlyDownward = () => {
    // setCanTrade(true);
    const prices = financialData.map((item) => item[1]);
    let n_current = 0;

    for (let i = 0; i < prices.length; i++) {
      if (prices[i] < prices[i + 1]) {
        n_current++;
        i++;
      }
    }
    if (n_current === 0) {
      // setCanTrade(false);
      return true;
    } else if (n_current > 0) {
      return false;
    }
  };

  //A pair of days: The day to buy and the day to sell (max price sell, min price buy)
  const handleFindSell = () => {
    setMaxVolume(0);
    setShowDown(false);
    setShowVolume(false);
    setDownwardSpan("");
    const minPrice = findMinPrice(financialData);
    setMinPriceDate(minPrice);
    const maxDate = findMaxPrice(financialData);
    setMaxPriceDate(maxDate);
  };

  //If the price only decreases in the date range, warn to not buy and not sell

  return (
    <div className="App">
      <div className="appParagraph">
        <p className="appTitle">Set the date range</p>
        <form onSubmit={(e) => handleSubmitAndFilter(e)}>
          <input
            className="formItem inputElement"
            type="date"
            placeholder="From"
            onInput={(e) => handleSetStart(e)}
          />
          <input
            className="formItem inputElement"
            type="date"
            placeholder="To"
            onInput={(e) => handleSetEnd(e)}
          />
          <button className="formItem submitButton" type="submit">
            Submit
          </button>
        </form>
      </div>

      <div className="appParagraph">
        {dataLoaded && financialData.length > 0 ? (
          <p style={{ color: "green", fontWeight: "700" }}>
            {" "}
            The data has loaded, choose what you want to know
          </p>
        ) : (
          <p> {loadMessage}</p>
        )}
        <button
          className="infoButton"
          onClick={() => {
            handleFindDownward();
          }}
        >
          Longest downward{" "}
        </button>

        <button
          className="infoButton"
          onClick={() => {
            handleFindVolume();
          }}
        >
          Highest trading volume
        </button>

        <button
          className="infoButton"
          onClick={() => {
            handleShouldBuyOrSell();
          }}
        >
          Sell and Buy
        </button>
      </div>

      <div className="appParagraph">
        {showVolume &&
          (maxVolume > 0 ? (
            <div>
              <p>
                Max volume: {maxVolume.toFixed(2)} euros on{" "}
                {convertToISO(maxVolumeDate)}
              </p>
            </div>
          ) : (
            <p>{message}</p>
          ))}

        {showTrade &&
          (canTrade ? (
            maxPriceDate ? (
              <div>
                {minPriceDate && (
                  <p>Should buy: on {convertToISO(minPriceDate)}</p>
                )}
                {maxPriceDate && (
                  <p>Should sell: on {convertToISO(maxPriceDate)}</p>
                )}
              </div>
            ) : (
              <p>{message}</p>
            )
          ) : (
            <p>Deals are not profitable: the price only decreases</p>
          ))}

        {showDown &&
          (downwardSpan ? (
            <div>
              <p>Longest downward span: {downwardSpan} days in a row</p>
            </div>
          ) : (
            <p>{message}</p>
          ))}
      </div>
    </div>
  );
}

export default App;
