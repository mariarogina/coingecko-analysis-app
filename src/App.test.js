import {
  findLongestDownward,
  findMaxVolume,
  findMaxPrice,
  findMinPrice,
  findDatesPrices,
} from "./utils";
import App from "./App";
import expect from "expect";
import renderer from "react-test-renderer";

//snapshot test
test("renders App component", () => {
  const componentTree = renderer.create(<App />).toJSON();
  expect(componentTree).toMatchSnapshot();
});

//unit test

const financialData = [
  [1, 2],
  [2, 3],
  [3, 2],
  [3, 4],
  [2, 1],
  [1, 2],
  [2, 3],
  [3, 4],
  [4, 5],
  [5, 6],
  [6, 7],
  [5, 4],
  [4, 3],
  [3, 2],
  [2, 1],
];
//unit test

const test_dates = [1, 2, 3, 3, 2, 1, 2, 3, 4, 5, 6, 5, 4, 3, 2];
const test_prices = [2, 3, 2, 4, 1, 2, 3, 4, 5, 6, 7, 4, 3, 2, 1];

describe("functions test", () => {
  it("finds longest downward", () => {
    const longestDown = findLongestDownward(financialData);
    expect(longestDown).toEqual(4);
  });
  it("finds dates", () => {
    const dates = findDatesPrices(financialData)[0];
    expect(dates.map((item) => item)).toEqual(test_dates.map((item) => item));
  });
  it("finds prices", () => {
    const prices = findDatesPrices(financialData)[1];
    expect(prices.map((item) => item)).toEqual(test_prices.map((item) => item));
  });
  it("finds max volume", () => {
    const maxData = findMaxVolume(financialData);
    expect(maxData[1]).toEqual(6);
  });
  it("finds max price", () => {
    const maxData = findMaxPrice(financialData);
    expect(maxData).toEqual(6);
  });
  it("finds min price", () => {
    const minData = findMinPrice(financialData);
    expect(minData).toEqual(2);
  });
});
