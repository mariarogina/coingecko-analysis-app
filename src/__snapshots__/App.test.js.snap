// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App component 1`] = `
<div
  className="App"
>
  <div
    className="appParagraph"
  >
    <h2
      style={
        Object {
          "color": "darkgreen",
        }
      }
    >
      Bitcoin Data Analyzer
    </h2>
    <p
      className="appTitle"
    >
      Set the date range
    </p>
    <form
      onSubmit={[Function]}
    >
      <input
        className="formItem inputElement"
        onInput={[Function]}
        placeholder="From"
        required={true}
        type="date"
      />
      <input
        className="formItem inputElement"
        onInput={[Function]}
        placeholder="To"
        required={true}
        type="date"
      />
      <button
        className="formItem submitButton"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
  <div
    className="appParagraph"
  >
    <p
      style={
        Object {
          "color": "red",
          "fontWeight": "700",
        }
      }
    >
       
      
    </p>
    <button
      className="infoButton"
      onClick={[Function]}
    >
      Longest downward
       
    </button>
    <button
      className="infoButton"
      onClick={[Function]}
    >
      Highest trading volume
    </button>
    <button
      className="infoButton"
      onClick={[Function]}
    >
      Sell and Buy
    </button>
  </div>
  <div
    className="appParagraph"
  />
</div>
`;
